src/XR/native/nativeXRFrame.ts

```diff
@@ -0,0 +1,92 @@
+import { RegisterNativeTypeAsync } from "../../Engines/nativeEngine";
+
+/** @hidden */
+interface INativeXRFrame extends XRFrame {
+    // Native-only helper functions
+    getPoseData: (space: XRSpace, baseSpace: XRReferenceSpace, vectorBuffer: ArrayBuffer, matrixBuffer: ArrayBuffer) => XRPose;
+}
+
+/** @hidden */
+export class NativeXRFrame implements XRFrame {
+    private readonly _xrTransform = new XRRigidTransform();
+    private readonly _xrPose: XRPose = {
+        transform: this._xrTransform,
+        emulatedPosition: false
+    };
+    // Enough space for position, orientation
+    private readonly _xrPoseVectorData = new Float32Array(4 * (4 + 4));
+
+    public get session(): XRSession {
+        return this._nativeImpl.session;
+    }
+
+    constructor(private _nativeImpl: INativeXRFrame) { }
+
+    public getPose(space: XRSpace, baseSpace: XRReferenceSpace): XRPose | undefined {
+        if (!this._nativeImpl.getPoseData(space, baseSpace, this._xrPoseVectorData.buffer, this._xrTransform.matrix.buffer)) {
+            return undefined;
+        }
+        const position = this._xrTransform.position as DOMPoint;
+        position.x = this._xrPoseVectorData[0];
+        position.y = this._xrPoseVectorData[1];
+        position.z = this._xrPoseVectorData[2];
+        position.w = this._xrPoseVectorData[3];
+
+        const orientation = this._xrTransform.orientation as DOMPoint;
+        orientation.x = this._xrPoseVectorData[4];
+        orientation.y = this._xrPoseVectorData[5];
+        orientation.z = this._xrPoseVectorData[6];
+        orientation.w = this._xrPoseVectorData[7];
+        return this._xrPose;
+    }
+
+    public fillPoses(spaces: XRSpace[], baseSpace: XRSpace, transforms: Float32Array): boolean {
+        return this._nativeImpl.fillPoses!(spaces, baseSpace, transforms);

```

Comment:
Why can functions in XRFrame be undefined? 🤔

Reply:
It's the spec's way of conveying optional support.  
I originally extended the interface to mark the functions as supported (not optional), but it seemed confusing to have the functions redefined with the one character missing:
https://github.com/BabylonJS/Babylon.js/pull/11655/commits/d93e90ddc786db0b5548dd2d80644e855f7dc4b8#diff-ab4efb00fe84c8b2b0443bfbe2b09f90255034e04359051f5fb41e1432f5060fR6-R9
I personally don't have much of a preference either way.